stdlib/public/core/UnsafeRawPointer.swift

```diff
@@ -1145,6 +1215,76 @@ extension UnsafeMutableRawPointer: Strideable {
   }
 }
 
+extension UnsafeMutableRawPointer {
+  /// Obtain the next pointer properly aligned to store a value of type `T`.
+  ///
+  /// If `self` is properly aligned for accessing `T`,
+  /// this function returns `self`.
+  ///
+  /// - Parameters:
+  ///   - type: the type to be stored at the returned address.
+  /// - Returns: a pointer properly aligned to store a value of type `T`.
+  @inlinable
+  @_alwaysEmitIntoClient
+  public func alignedUp<T>(for type: T.Type) -> Self {
+    let mask = UInt(Builtin.alignof(T.self)) &- 1
+    let bits = (UInt(Builtin.ptrtoint_Word(_rawValue)) &+ mask) & ~mask
+    return .init(Builtin.inttoptr_Word(bits._builtinWordValue))
+  }
+
+  /// Obtain the preceding pointer properly aligned to store a value of type `T`.
+  ///
+  /// If `self` is properly aligned for accessing `T`,
+  /// this function returns `self`.
+  ///
+  /// - Parameters:
+  ///   - type: the type to be stored at the returned address.
+  /// - Returns: a pointer properly aligned to store a value of type `T`.
+  @inlinable
+  @_alwaysEmitIntoClient
+  public func alignedDown<T>(for type: T.Type) -> Self {
+    let mask = UInt(Builtin.alignof(T.self)) &- 1
+    let bits = UInt(Builtin.ptrtoint_Word(_rawValue)) & ~mask
+    return .init(Builtin.inttoptr_Word(bits._builtinWordValue))
+  }
+
+  /// Obtain the next pointer whose bit pattern is a multiple of `alignment`.
+  ///
+  /// If the bit pattern of `self` is a multiple of `alignment`,
+  /// this function returns `self`.
+  ///
+  /// - Parameters:
+  ///   - alignment: the alignment of the returned pointer, in bytes.
+  ///     `alignment` must be a whole power of 2.
+  /// - Returns: a pointer aligned to `alignment`.
+  @inlinable
+  @_alwaysEmitIntoClient
+  public func alignedUp(toMultipleOf alignment: Int) -> Self {
+    let mask = alignment &- 1
+    _debugPrecondition(alignment & mask == 0)
+    let bits = (Int(Builtin.ptrtoint_Word(_rawValue)) &+ mask) & ~mask
```

Comment:
It's not an issue on most platforms, because addresses very near near the top of the address space are usually reserved and not valid pointers, but I think that formally we want this to be a non-wrapping addition on `UInt` to prevent aligning up from wrapping around to address zero.